pilot/pkg/security/authz/builder/builder.go

```diff
@@ -143,30 +190,83 @@ func build(policies []model.AuthorizationPolicy, tdBundle trustdomain.Bundle,
 		}
 	}
 
-	return &rbachttppb.RBAC{Rules: rules}
+	if forTCP {
+		return &builtConfigs{tcp: b.buildTCP(rules, providers)}
+	}
+	return &builtConfigs{http: b.buildHTTP(rules, providers)}
 }
 
-// nolint: interfacer
-func createHTTPFilter(config *rbachttppb.RBAC) *httppb.HttpFilter {
-	if config == nil {
-		return nil
+func (b Builder) buildHTTP(rules *rbacpb.RBAC, providers []string) []*httppb.HttpFilter {
+	if !b.option.IsCustomBuilder {
+		rbac := &rbachttppb.RBAC{Rules: rules}
+		return []*httppb.HttpFilter{
+			{
+				Name:       authzmodel.RBACHTTPFilterName,
+				ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
+			},
+		}
 	}
-	return &httppb.HttpFilter{
-		Name:       authzmodel.RBACHTTPFilterName,
-		ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(config)},
+
+	extauthz, err := buildExtAuthz(b.extensions, providers)
+	if err != nil {
+		authzLog.Errorf("failed parsing CUSTOM action, will generate a deny all config: %v", err)
+		rbac := &rbachttppb.RBAC{Rules: rbacDefaultDenyAll}
+		return []*httppb.HttpFilter{
+			{
+				Name:       authzmodel.RBACHTTPFilterName,
+				ConfigType: &httppb.HttpFilter_TypedConfig{TypedConfig: util.MessageToAny(rbac)},
+			},
+		}
+	}
+	// Add the RBAC filter in shadow mode so that it only evaluates the matching rules for CUSTOM action but not enforce it.
```

Comment:
I read the sample output and code and here is my understanding: 

- we generate additional rbac filter, with shadow deny policies. Those rbac filter's condition is based on ext_authz condition.
- Then the ext authz filter is using the preivous rbac filte's output metadata filter, with `filterEnabledMetadata` to determine whether acutal kick in.

In some way, the first rbac is only to prepapre the triggering condition for the ext authz's config limitation?

Is that correct? Feel free to point to some section in the design doc. I do feel though, this is worth a bit more detailed comments around the code, maybe here and other places.

Reply:
that's correct, I think this is already pretty clear with the current comment, added a link to the design doc for more details.

Reply:
Where's the link of the design doc? didn't see it nearby here.

We can add one sentence like "...keyed by the policy. And then the ext_authz filter can utilize these metadata to trigger the enforcement conditionally."

optional.

Reply:
updated the comment, the design link is just below.